README.md
# Wikidata Query Service (WDQS) Dashboard - ARCHIVED

This codebase is part of the legacy [Discovery Dashboards](https://discovery.wmflabs.org/) project. It is kept for archival purposes. It was maintained by [Mikhail Popov](https://meta.wikimedia.org/wiki%2fUser%3aMPopov_%28WMF%29).

## Quick start

Install the dependencies:

```R
install.packages(c("remotes", "shiny", "reshape2", "data.table"))
remotes::install_git("https://gerrit.wikimedia.org/r/wikimedia/discovery/polloi")
```

Run the app:

```R
shiny::runApp(launch.browser = 0)
```

Please note that this project is licensed under [MIT License](LICENSE.md) and released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms. [Wikimedia technical spaces code of conduct](https://www.mediawiki.org/wiki/Special:MyLanguage/Code_of_Conduct) also applies.


CHANGELOG.md
# Change Log (Patch Notes)

All notable changes to this project will be documented in this file.

## 2021/10/01
- Added decommission notice ([T227782](https://phabricator.wikimedia.org/T227782))

## 2017/06/20
- Added licensing info ([T167930](https://phabricator.wikimedia.org/T167930))

## 2017/02/02
- Updated to work with new datasets generated by Reportupdater-based golden ([T150915](https://phabricator.wikimedia.org/T150915))
- Added LDF endpoint usage ([T153936](https://phabricator.wikimedia.org/T153936))

## 2015/11/10
- Refactored large sections of code
- Added smoothing and time frame selection

## 2015/09/30
- Added a change log
- Added a contributor code of conduct
- Updated the readme


ui.R
library(shiny)
library(shinydashboard)
library(dygraphs)

spider_checkbox <- function(input_id) {
  shiny::checkboxInput(input_id, "Include automata", value = TRUE, width = NULL)
}

function(request) {
  dashboardPage(

    dashboardHeader(title = "Wikidata Query Service", dropdownMenuOutput("message_menu"), disable = FALSE),

    dashboardSidebar(
      tags$head(
        tags$link(rel = "stylesheet", type = "text/css", href = "custom.css"),
        tags$script(src = "custom.js")
      ),
      sidebarMenu(id = "tabs",
                  menuItem(text = "Endpoint Usage", tabName = "endpoint_usage"),
                  menuItem(text = "Homepage Traffic", tabName = "wdqs_visits"),
                  menuItem(text = "Global Settings",
                           selectInput(inputId = "smoothing_global", label = "Smoothing", selectize = TRUE, selected = "day",
                                       choices = c("No Smoothing" = "day", "Weekly Median" = "week", "Monthly Median" = "month", "Splines" = "gam")),
                           icon = icon("cog", lib = "glyphicon"))
      ),
      div(icon("info-sign", lib = "glyphicon"), HTML("<strong>Tip</strong>: you can drag on the graphs with your mouse to zoom in on a particular date range."), style = "padding: 10px; color: white;"),
      div(bookmarkButton(), style = "text-align: center;")
    ),

    dashboardBody(
      tabItems(
        tabItem(tabName = "endpoint_usage",
                includeHTML("./tab_documentation/notice.html"),
                fluidRow(
                  column(polloi::smooth_select("smoothing_usage"), width = 4),
                  column(checkboxInput("usage_logscale", "Use Log scale", TRUE), width = 4),
                  column(spider_checkbox("include_automata_usage"), width = 4)),
                dygraphOutput("sparql_usage_plot", height = "200px"),
                dygraphOutput("ldf_usage_plot", height = "200px"),
                fluidRow(div(id = "usage_legend"), style = "padding-top: 10px; height: 20px; text-align: center;"),
                includeMarkdown("./tab_documentation/wdqs_usage.md")),
        tabItem(tabName = "wdqs_visits",
                includeHTML("./tab_documentation/notice.html"),
                fluidRow(
                  column(polloi::smooth_select("smoothing_visits"), width = 4),
                  column(checkboxInput("visits_logscale", "Use Log scale", TRUE), width = 4),
                  column(spider_checkbox("include_automata_visits"), width = 4)),
                dygraphOutput("wdqs_visits_plot", height = "400px"),
                fluidRow(div(id = "wdqs_visits_legend"), style = "padding-top: 10px; height: 20px; text-align: center;"),
                includeMarkdown("./tab_documentation/wdqs_visits.md"))
      )
    ),

    skin = "purple", title = "Wikidata Query Service Usage Dashboard | Search Platform | Technology | Wikimedia Foundation")
}


utils.R
library(magrittr)

read_wdqs <- function() {
  data <- polloi::read_dataset("discovery/metrics/wdqs/basic_usage.tsv", col_types = "Dclli") %>%
    dplyr::arrange(date)
  wdqs_visits <<- data %>%
    dplyr::filter(path == "/" & http_success) %>%
    dplyr::select(c(date, is_automata, requests = events))
  sparql_usage <<- data %>%
    dplyr::filter(path == "/bigdata/namespace/wdq/sparql" & http_success) %>%
    dplyr::select(c(date, is_automata, requests = events))
  ldf_usage <<- data %>%
    dplyr::filter(path == "/bigdata/ldf" & http_success) %>%
    dplyr::select(c(date, is_automata, requests = events))
  return(invisible())
}

spider_subset <- function(data, val) {
  if (!val) {
    data <- data[!data$is_automata,]
  }
  return(dplyr::summarize(dplyr::group_by(data, date), requests = sum(requests)))
}

conditional_transform <- function(x, cond, .f, ...) {
  if (cond) {
    return(data_transform(x, .f, ...))
  } else {
    return(x)
  }
}

data_transform <- function(x, .f = identity, ...) {
  col_names <- setdiff(colnames(x), "date")
  x[, col_names] <- .f(x[, col_names], ...)
  return(x)
}

pow <- function(base, exponent) {
  return(base^exponent)
}

exp10 <- function(x) {
  return(pow(10, x))
}


data_retrieval/wdqs.R
#Config variables and setup
options(scipen = 500,
        q = "no")
base_path <- "/a/aggregate-datasets/wdqs/"

#Dependencies
library(olivr)
suppressPackageStartupMessages(library(data.table))

if(!file.exists(base_path)){
  dir.create(path = base_path)
}

#Conditional write; if the file exists, add x to the end. If it doesn't, write an entirely new file.
conditional_write <- function(x, file) {
  if(file.exists(file)){
    write.table(x, file, append = TRUE, sep = "\t", row.names = FALSE, col.names = FALSE)
  } else {
    write.table(x, file, append = FALSE, sep = "\t", row.names = FALSE)
  }
}

#Retrieves data for the WDQS stuff we care about, drops it in the aggregate-datasets directory.
#Should be run on stat1002, /not/ on the datavis machine.

# Create a script that would produce raw data on usage of
# - query.wikidata.org
# - SPARQL endpoint: query.wikidata.org/bigdata/namespace/wdq/sparql

#A function for creating a hive query aimed at the data from yesterday.
# Central function
main <- function(date = NULL) {

  # Date handling
  if(is.null(date)) {
    date <- Sys.Date() - 1
  }
  subquery <- paste0(" WHERE year = ", lubridate::year(date),
                     " AND month = ", lubridate::month(date),
                     " AND day = ", lubridate::day(date), " ")

  # Write query and dump to file
  query <- paste0("USE wmf;
                   SELECT year, month, day,
                   FIND_IN_SET(uri_path, '/bigdata/namespace/wdq/sparql,/,/index.php') AS uri_path,
                   IF(INSTR(uri_query, 'query') > 0, 'query', 'other') AS uri_query,
                   IF(INSTR(content_type, 'sparql-results') > 0, 'sparql results', 'other') AS content_type,
                   COUNT(*) AS n
                   FROM webrequest",
                   subquery,
                  "AND webrequest_source = 'misc' AND FIND_IN_SET(http_status, '200,301,302,303') > 0
                   GROUP BY year, month, day,
                   FIND_IN_SET(uri_path, '/bigdata/namespace/wdq/sparql,/,/index.php'),
                   IF(INSTR(uri_query, 'query') > 0, 'query', 'other'),
                   IF(INSTR(content_type, 'sparql-results') > 0, 'sparql results', 'other');")
  query_dump <- tempfile()
  cat(query, file = query_dump)

  # Query
  results_dump <- tempfile()
  system(paste0("export HADOOP_HEAPSIZE=1024 && hive -f ", query_dump, " > ", results_dump))
  results <- read.delim(results_dump, sep = "\t", quote = "", as.is = TRUE, header = TRUE)
  file.remove(query_dump, results_dump)

  results$uri_path <- factor(results$uri_path, 0:3, c("other", "/bigdata/namespace/wdq/sparql", "/", "/index.php"))

  output <- data.frame(timestamp = as.Date(paste(results$year, results$month, results$day, sep = "-")),
                       path = results$uri_path,
                       query = results$uri_query,
                       content = results$content_type,
                       events = results$n,
                       stringsAsFactors = FALSE)

  # Write out
  conditional_write(output, file.path(base_path, "wdqs_aggregates.tsv"))

}

# Backlog (start date: 2015-07-28):
# backlog <- function(days) {
#   for (i in days:1) try(main(Sys.Date() - i), silent = TRUE)
# }; backlog(30) # as of 2015-08-27

#Run and kill
main()
q(save = "no")


server.R
library(shiny)
library(shinydashboard)
library(dygraphs)

source("utils.R")

existing_date <- Sys.Date() - 1

shinyServer(function(input, output, session) {

  if(Sys.Date() != existing_date){
    read_wdqs()
    existing_date <<- Sys.Date()
  }

  output$ldf_usage_plot <- renderDygraph(
    ldf_usage %>%
      spider_subset(val = input$include_automata_usage) %>%
      # The next few lines make for better smoothing because the data is first log-transformed:
      conditional_transform(input$usage_logscale && polloi::smooth_switch(input$smoothing_global, input$smoothing_usage) != "day", log10) %>%
      # ...THEN smoothed:
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_usage), rename = FALSE) %>%
      # ...and then exp10-transformed back to the original scale:
      conditional_transform(input$usage_logscale && polloi::smooth_switch(input$smoothing_global, input$smoothing_usage) != "day", exp10) %>%
      polloi::make_dygraph(xlab = "Date", ylab = "Requests", title = "Daily LDF usage", group = "usage") %>%
      # ...because we're using dygraphs' native log-scaling:
      dyAxis("y", logscale = input$usage_logscale) %>%
      dyLegend(labelsDiv = "usage_legend") %>%
      dyRangeSelector %>%
      dyEvent(as.Date("2017-01-01"), "D (Started tracking LDF usage)", labelLoc = "bottom")
  )

  output$sparql_usage_plot <- renderDygraph(
    sparql_usage %>%
      spider_subset(val = input$include_automata_usage) %>%
      # See above for why we're conditional_transform'ing here.
      conditional_transform(input$usage_logscale && polloi::smooth_switch(input$smoothing_global, input$smoothing_usage) != "day", log10) %>%
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_usage), rename = FALSE) %>%
      conditional_transform(input$usage_logscale && polloi::smooth_switch(input$smoothing_global, input$smoothing_usage) != "day", exp10) %>%
      polloi::make_dygraph(xlab = "Date", ylab = "Requests", title = "Daily SPARQL usage", group = "usage") %>%
      dyLegend(labelsDiv = "usage_legend") %>%
      dyAxis("y", logscale = input$usage_logscale) %>%
      dyRangeSelector %>%
      dyEvent(as.Date("2015-09-07"), "A (Announcement)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2015-11-05"), "B (Labs bot)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2016-12-28"), "C (Bot ruleset)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-02-18"), "E (Alias)", labelLoc = "bottom")
  )

  output$wdqs_visits_plot <- renderDygraph(
    wdqs_visits %>%
      spider_subset(val = input$include_automata_visits) %>%
      # The next few lines make for better smoothing because the data is first log-transformed:
      conditional_transform(input$visits_logscale && polloi::smooth_switch(input$smoothing_global, input$smoothing_visits) != "day", log10) %>%
      # ...THEN smoothed:
      polloi::smoother(smooth_level = polloi::smooth_switch(input$smoothing_global, input$smoothing_visits), rename = FALSE) %>%
      # ...and then exp10-transformed back to the original scale:
      conditional_transform(input$visits_logscale && polloi::smooth_switch(input$smoothing_global, input$smoothing_visits) != "day", exp10) %>%
      polloi::make_dygraph(xlab = "Date", ylab = "Pageviews", title = "Daily WDQS Homepage Traffic") %>%
      # ...because we're using dygraphs' native log-scaling:
      dyAxis("y", logscale = input$visits_logscale) %>%
      dyLegend(labelsDiv = "wdqs_visits_legend") %>%
      dyEvent(as.Date("2015-09-07"), "A (Announcement)", labelLoc = "bottom") %>%
      dyEvent(as.Date("2017-01-01"), "R (reportupdater)", labelLoc = "bottom")
  )

  # Check datasets for missing data and notify user which datasets are missing data (if any)
  output$message_menu <- renderMenu({
    notifications <- list(
      polloi::check_yesterday(sparql_usage, "WDQS usage data"),
      polloi::check_past_week(sparql_usage, "WDQS usage data"))
    notifications <- notifications[!sapply(notifications, is.null)]
    return(dropdownMenu(type = "notifications", .list = notifications))
  })

})
